presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -602,13 +607,30 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<RowExpression> context)
                 output = new FilterNode(node.getSourceLocation(), idAllocator.getNextId(), output, postJoinPredicate);
             }
 
-            if (!node.getOutputVariables().equals(output.getOutputVariables())) {
```

Comment:
can we instead keep this but also add the condition node.isCrossJoin()?  Then we could keep the check in joinNode and wouldn't need the extra rule to add a projection over cross joins.

Reply:
Okay made the change, but will make an issue to address/discuss this more